text/3806-align-attr.md

```diff
@@ -0,0 +1,541 @@
+- Feature Name: `align_attr`
+- Start Date: 2025-05-01
+- RFC PR: [rust-lang/rfcs#3806](https://github.com/rust-lang/rfcs/pull/3806)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add an `#[align(…)]` attribute to set the minimum alignment of `struct` and
+`enum` fields, `static`s, functions, and local variables.
+
+# Motivation
+[motivation]: #motivation
+
+## Bindings to C and C++
+
+[C](https://en.cppreference.com/w/c/language/_Alignas) and
+[C++](https://en.cppreference.com/w/cpp/language/alignas) provide an `alignas`
+modifier to set the alignment of specific struct fields. To represent such
+structures in Rust, `bindgen` is sometimes forced to add explicit padding
+fields:
+
+```c
+// C code
+#include <stdint.h>
+struct foo {
+    uint8_t x;
+    _Alignas(128) uint8_t y;
+    uint8_t z;
+};
+```
+
+```rust
+// Rust bindings generated by `bindgen`
+#[repr(C, align(128))]
+pub struct foo {
+    pub x: u8,
+    pub __bindgen_padding_0: [u8; 127usize],
+    pub y: u8,
+    pub z: u8,
+}
+```
+
+The `__bindgen_padding_0` field makes the generated bindings more confusing and
+less ergonomic. Also, it is unsound: the padding should be using `MaybeUninit`.
+And even then, there is no guarantee of ABI compatibility on all potential
+targets.
+
+## Packing values into fewer cache lines
+
+When working with large values (lookup tables, for example), it is often
+desirable, for optimal performance, to pack them into as few cache lines as
+possible. One way of doing this is to force the alignment of the value to be at
+least the size of the cache line, or perhaps the greatest common denominator of
+the value and cache line sizes.
```

Comment:
I am very confused by this paragraph, and don't understand what it tries to say. By raising the alignment, the items will be further apart from each other (since each one must be properly aligned), and therefore it will take *more* cache lines, not fewer?

Reply:
The idea is that you don't want a field to span across 2 cache lines if it could fit into just one. An easy way to achieve that is to make the field start at the start of a new cache line. 

Aligning a field also means that a reference to it is aligned, which can improve performance in certain cases.

Reply:
I’ve reworded this paragraph.

Reply:
Right, so you want to avoid false sharing, that makes sense.

The new text helps, thanks. The section title still says "Packing values into fewer cache lines" though, which doesn't seem to match the text.

Reply:
It's not about false sharing -- if that was a problem, trailing padding until the next cache line boundary would be a plus. It's about possibly bringing fewer cache lines into the cache for common access patterns. This is a minor micro-optimization at best.

Reply:
> It's about possibly bringing fewer cache lines into the cache for common access patterns.

This is exactly it.

Reply:
> "Packing values into fewer cache lines"

As it pertains to the header, the ambiguity is the difference between packing the set of all values into (maybe) fewer cache lines, or packing each value into (maybe) fewer cache lines.